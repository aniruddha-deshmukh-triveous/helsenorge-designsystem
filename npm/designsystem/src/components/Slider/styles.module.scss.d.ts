export type Styles = {
  slider: string;
  slider__options: string;
  slider__text: string;
  slider__track: string;
  'slider__track--disabled': string;
  'slider__track-wrapper': string;
  'slider__track-wrapper--disabled': string;
  slider__trigger: string;
  'slider__trigger--disabled': string;
  'slider__trigger-inner': string;
  'slider__trigger-inner--disabled': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
