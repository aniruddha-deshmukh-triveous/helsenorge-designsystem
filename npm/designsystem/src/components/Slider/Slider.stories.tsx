import React from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import Slider from './Slider';
import { withA11y } from '@storybook/addon-a11y';

const stories = storiesOf('Slider', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <div style={{ width: '30rem' }}>
    <Slider title={'Steps: 100'} labelLeft={'left'} labelRight={'right'} step={1} />
    <Slider title={'Steps: 10'} step={10} />
    <Slider title={'Steps: 3'} step={50} />
    <Slider title={'Disabled'} disabled={true} />
  </div>
));
