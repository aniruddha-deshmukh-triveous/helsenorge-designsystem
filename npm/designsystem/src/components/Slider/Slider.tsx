import React, { useEffect, useState, useRef } from 'react';
import classNames from 'classnames';

import Title from '../Title';

import {
  stopEvent,
  getMousePosition,
  getElementWidth,
  notifyMove,
  calculateSliderPositionBasedOnValue,
  calculateValueBasedOnSliderPosition,
  calculateChangeOfPosition,
  calculateSliderTranslate,
  addMouseListeners,
  removeMouseListeners,
  addTouchListeners,
  removeTouchListeners,
} from './SliderUtils';

import SliderStyles from './styles.module.scss';

interface SliderProps {
  /** Tekst som vises til over feltet */
  title?: string;
  /** Tekst som vises til venstre i feltet */
  labelLeft?: string;
  /** Tekst som vises til høyre i feltet */
  labelRight?: string;
  /** Step value som brukes når verdien endres på slider'en */
  step?: number;
  /** Om feltet er disabled */
  disabled?: boolean;
  /** Function som kalles når verdien i feltet endres */
  onChange?: (value: number) => void;
}

export const Slider = React.forwardRef(function SliderForwardedRef(props: SliderProps, ref: React.ForwardedRef<HTMLElement>) {
  const { title, labelLeft, labelRight, disabled = false, step = 1, onChange } = props;
  const [value, setValue] = useState(50);
  const [isMouseDown, setIsMouseDown] = useState(false);
  const [trackerWidth, setTrackerWidth] = useState(0);
  const [sliderWidth, setSliderWidth] = useState(0);
  const [sliderXPos, setSliderXPos] = useState(0);
  const [sliderTemporaryXPos, setSliderTemporaryXPos] = useState(0);
  const trackerRef = ref ? (ref as React.RefObject<HTMLDivElement>) : useRef<HTMLDivElement>(null);
  const sliderRef = useRef<HTMLDivElement>(null);
  const min = 0;
  const max = 100;
  const moveMouseEvent = (evt: MouseEvent) => {
    onMouseMove(evt);
  };
  const mouseUpEvent = (evt: MouseEvent) => {
    onMouseUp(evt);
  };
  const moveTouchEvent = (evt: TouchEvent) => {
    onMouseMove(evt);
  };
  const touchUpEvent = (evt: TouchEvent) => {
    onMouseUp(evt);
  };

  useEffect(() => {
    const trackerWidth = getElementWidth(trackerRef.current);
    const sliderWidth = getElementWidth(sliderRef.current);
    setSliderXPos(calculateSliderPositionBasedOnValue(value, trackerWidth, sliderWidth, max, min));
    setTrackerWidth(trackerWidth);
    setSliderWidth(sliderWidth);
  }, []);

  useEffect(() => {
    if (isMouseDown) {
      addMouseListeners(moveMouseEvent, mouseUpEvent);
      addTouchListeners(moveTouchEvent, touchUpEvent);
    }
  }, [isMouseDown]);

  useEffect(() => {
    if (!disabled) {
      notifyMove(value, onChange);
    }
  }, [value]);

  const onKeyDown = (evt: React.KeyboardEvent<HTMLDivElement>) => {
    if (disabled) return;
    const size: number = max - min;
    const pixelPerSize: number = (trackerWidth - sliderWidth) / size;
    if (evt.key === 'ArrowRight' || evt.key === 'ArrowDown') {
      updateSliderPosition(step * pixelPerSize);
    }
    if (evt.key === 'ArrowLeft' || evt.key === 'ArrowUp') {
      updateSliderPosition(-step * pixelPerSize);
    }
  };

  const onMouseOrTouchDown = (e: MouseEvent | React.MouseEvent<{}> | TouchEvent | React.TouchEvent<{}>): void => {
    if (disabled) return;
    const trackerWidth = getElementWidth(trackerRef.current);
    const sliderWidth = getElementWidth(sliderRef.current);
    const updatedMousePosition: number = getMousePosition(e);
    calculateSliderTranslate(updatedMousePosition, sliderRef.current, sliderWidth, updateSliderPosition);
    setIsMouseDown(true);
    setSliderTemporaryXPos(updatedMousePosition);
    setTrackerWidth(trackerWidth);
    setSliderWidth(sliderWidth);
    stopEvent(e);
  };

  const onMouseUp = (e: MouseEvent | React.MouseEvent<{}> | TouchEvent | React.TouchEvent<{}>): void => {
    if (disabled) return;
    setIsMouseDown(false);
    removeMouseListeners(moveMouseEvent, mouseUpEvent);
    removeTouchListeners(moveTouchEvent, touchUpEvent);
    stopEvent(e);
  };

  const onMouseMove = (e: MouseEvent | React.MouseEvent<{}> | TouchEvent | React.TouchEvent<{}>): void => {
    if (disabled || !isMouseDown) return;
    const updatedMousePosition: number = getMousePosition(e);
    const diff: number = updatedMousePosition - sliderTemporaryXPos;
    updateSliderPosition(diff);
    stopEvent(e);
  };

  const updateSliderPosition = (diff: number): void => {
    if (diff === 0) return;
    const updatedSliderPos: number = calculateChangeOfPosition(diff, sliderXPos, trackerWidth, sliderWidth);
    setSliderXPos(updatedSliderPos);
    setSliderTemporaryXPos(updatedSliderPos + diff);
    setValue(calculateValueBasedOnSliderPosition(updatedSliderPos, max, min, trackerWidth, sliderWidth, step));
  };

  return (
    <div className={SliderStyles.slider}>
      {title && (
        <Title htmlMarkup={'h3'} margin={1.5} appearance={'title3'}>
          {title}
        </Title>
      )}
      <div
        ref={trackerRef}
        className={classNames(SliderStyles['slider__track-wrapper'], disabled ? SliderStyles['slider__track-wrapper--disabled'] : '')}
        onMouseDown={onMouseOrTouchDown}
        onTouchStart={onMouseOrTouchDown}
        data-testid={'tracker'}
      >
        <div className={classNames(SliderStyles.slider__track, disabled ? SliderStyles['slider__track--disabled'] : '')} />
        <div
          ref={sliderRef}
          className={classNames(SliderStyles.slider__trigger, disabled ? SliderStyles['slider__trigger--disabled'] : '')}
          style={{
            left: `${sliderXPos}px`,
          }}
          onKeyDown={onKeyDown}
          role={'slider'}
          aria-valuenow={value}
          aria-valuemin={min}
          aria-valuemax={max}
          aria-label={`${title ? title + ' ' : ''}${labelLeft ? labelLeft + ' ' : ''}${labelRight ? labelRight + ' ' : ''}`}
          tabIndex={disabled ? -1 : 0}
        >
          <div
            className={classNames(SliderStyles['slider__trigger-inner'], disabled ? SliderStyles['slider__trigger-inner--disabled'] : '')}
          />
        </div>
      </div>
      {(labelLeft || labelRight) && (
        <span className={SliderStyles.slider__options}>
          <p className={SliderStyles.slider__text} aria-hidden={true}>
            {labelLeft}
          </p>
          <p className={SliderStyles.slider__text} aria-hidden={true}>
            {labelRight}
          </p>
        </span>
      )}
    </div>
  );
});

export default Slider;
