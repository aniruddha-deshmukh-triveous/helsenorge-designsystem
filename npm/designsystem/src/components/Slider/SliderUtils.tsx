export const stopEvent = (e: MouseEvent | React.MouseEvent<{}> | TouchEvent | React.TouchEvent<{}>): boolean => {
  if (e.stopPropagation) {
    e.stopPropagation();
  }
  if (e.preventDefault) {
    e.preventDefault();
  }
  return false;
};

export const getMousePosition = (e: MouseEvent | React.MouseEvent<{}> | TouchEvent | React.TouchEvent<{}>): number => {
  if (isTouchEvent(e)) {
    const touch: Touch = (e as TouchEvent).touches[0];
    return touch.pageX;
  }
  return (e as MouseEvent).pageX;
};

export const isTouchEvent = (e: MouseEvent | React.MouseEvent<{}> | TouchEvent | React.TouchEvent<{}>): boolean => {
  if (
    e.type === 'touchcancel' ||
    e.type === 'touchend' ||
    e.type === 'touchenter' ||
    e.type === 'touchleave' ||
    e.type === 'touchmove' ||
    e.type === 'touchstart'
  ) {
    if ((e as TouchEvent).touches.length === 0) {
      return false;
    } else {
      return true;
    }
  }
  return false;
};

export const notifyMove = (value: number, onChange?: (value: number) => void): void => {
  if (onChange) {
    onChange(value);
  }
};

export const calculateSliderPositionBasedOnValue = (
  value: number,
  trackWidth: number,
  sliderWidth: number,
  max: number,
  min: number
): number => {
  const size: number = max - min;
  const pixelPerSize: number = (trackWidth - sliderWidth) / size;
  return pixelPerSize * value;
};

export const calculateValueBasedOnSliderPosition = (
  sliderPosition: number,
  max: number,
  min: number,
  trackerWidth: number,
  sliderWidth: number,
  step: number
): number => {
  const size: number = max - min;
  const pixelPerSize: number = (trackerWidth - sliderWidth) / size;
  const newValue: number = Math.round(sliderPosition / pixelPerSize);
  return alignValue(newValue, step, max);
};

export const alignValue = (value: number, step: number, max: number): number => {
  const valModStep: number = value % step;
  let alignedValue: number = value - valModStep;

  const possibleMaxValueWithCurrentStep: number = Math.floor(max / step) * step;
  if (value > possibleMaxValueWithCurrentStep) {
    const diff: number = max - possibleMaxValueWithCurrentStep;
    if (value > possibleMaxValueWithCurrentStep + diff / 2) {
      alignedValue = max;
    }
  }

  return Math.round(alignedValue);
};

export const addMouseListeners = (moveMouseEvent: any, mouseUpEvent: any): void => {
  document.addEventListener('mousemove', moveMouseEvent, false);
  document.addEventListener('mouseup', mouseUpEvent, false);
};

export const removeMouseListeners = (moveMouseEvent: any, mouseUpEvent: any): void => {
  document.removeEventListener('mousemove', moveMouseEvent, false);
  document.removeEventListener('mouseup', mouseUpEvent, false);
};

export const addTouchListeners = (moveTouchEvent: any, touchUpEvent: any): void => {
  document.addEventListener('touchmove', moveTouchEvent, false);
  document.addEventListener('touchend', touchUpEvent, false);
};

export const removeTouchListeners = (moveTouchEvent: any, touchUpEvent: any): void => {
  document.removeEventListener('touchmove', moveTouchEvent, false);
  document.removeEventListener('touchend', touchUpEvent, false);
};
