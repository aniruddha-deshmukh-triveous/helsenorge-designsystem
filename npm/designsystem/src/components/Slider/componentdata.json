{
  "props": {
    "title": {
      "defaultValue": null,
      "description": "Sets the title of the slider.",
      "name": "title",
      "parent": {
        "fileName": "src/components/Slider/Slider.tsx",
        "name": "SliderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Slider/Slider.tsx",
          "name": "SliderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "labelLeft": {
      "defaultValue": null,
      "description": "Adds the left hand label to the element.",
      "name": "labelLeft",
      "parent": {
        "fileName": "src/components/Slider/Slider.tsx",
        "name": "SliderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Slider/Slider.tsx",
          "name": "SliderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "labelRight": {
      "defaultValue": null,
      "description": "Adds the right hand label to the element.",
      "name": "labelRight",
      "parent": {
        "fileName": "src/components/Slider/Slider.tsx",
        "name": "SliderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Slider/Slider.tsx",
          "name": "SliderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "step": {
      "defaultValue": null,
      "description": "Decides the number of steps for each movement of the slider.",
      "name": "step",
      "parent": {
        "fileName": "src/components/Slider/Slider.tsx",
        "name": "SliderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Slider/Slider.tsx",
          "name": "SliderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "number"
      }
    },
    "disabled": {
      "defaultValue": null,
      "description": "Disables the slider element.",
      "name": "disabled",
      "parent": {
        "fileName": "src/components/Slider/Slider.tsx",
        "name": "SliderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Slider/Slider.tsx",
          "name": "SliderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "onChange": {
      "defaultValue": null,
      "description": "Function to be called when the value state has changed.",
      "name": "onChange",
      "parent": {
        "fileName": "src/components/Slider/Slider.tsx",
        "name": "SliderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Slider/Slider.tsx",
          "name": "SliderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "((value: number) => void)"
      }
    }
  }
}
