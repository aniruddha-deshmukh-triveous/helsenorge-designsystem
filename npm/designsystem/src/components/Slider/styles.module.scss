@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';

.slider {
  width: 100%;
  padding: getSpacer(m) 0;
  position: relative;
  outline: none !important;

  &__track-wrapper {
    width: 100%;
    cursor: pointer;
    touch-action: none;

    &--disabled {
      cursor: default;
    }
  }

  &__track {
    border-top: 2px solid $black;
    width: 100%;
    position: relative;
    top: 1rem;

    &--disabled {
      border-color: $neutral600;
    }
  }

  &__marker {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 2rem;
    height: 2rem;
    background-color: $white;
    border: 0.15rem solid $black;
    border-radius: 1rem;
    position: relative;
    top: -0.05rem;
    z-index: 9;
    outline: none;
    cursor: pointer;

    &:focus-within {
      box-shadow: 0 0 0 0.1rem $black;
    }

    &::after {
      content: '';
      width: 1.25rem;
      height: 1.25rem;
      border: 0.1rem solid $black;
      border-radius: 1rem;
      background-color: $blueberry600;

      :hover > &,
      :focus > & {
        background-color: $blueberry700;
      }
    }

    &--disabled {
      border-color: $neutral500;
      cursor: default;

      &::after {
        border-color: $neutral500;
        background-color: $neutral200;

        :hover > &,
        :focus > & {
          background-color: $neutral200;
        }
      }
    }
  }

  &__options {
    display: inline-flex;
    justify-content: space-between;
    width: 100%;
  }
}
