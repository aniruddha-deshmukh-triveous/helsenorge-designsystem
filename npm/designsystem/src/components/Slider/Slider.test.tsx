import React from 'react';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { Slider } from './Slider';

describe('Gitt at Slider skal vises', (): void => {
  describe('Når slider rendres', (): void => {
    test('Så vises slider riktig', (): void => {
      const title = 'This is a title';
      const optionLeft = 'left';
      const optionRight = 'right';
      const { container } = render(<Slider title={title} labelLeft={optionLeft} labelRight={optionRight} />);
      expect(container).toMatchSnapshot();

      const titleElement = screen.getByText(title);
      expect(titleElement).toBeTruthy();

      const optionLeftElement = screen.getByText(optionLeft);
      expect(optionLeftElement).toBeTruthy();

      const optionRightElement = screen.getByText(optionRight);
      expect(optionRightElement).toBeTruthy();

      const sliderElement = screen.getByRole('slider');
      expect(sliderElement.className).toBe('slider__trigger');
    });

    test('Så er sliderElement posisjonert riktig', () => {
      const original = global.document['window'];
      const getBoundingClientRect = jest
        .fn()
        .mockReturnValueOnce({ left: 100, right: 500 })
        .mockReturnValueOnce({ left: 150, right: 450 });
      window.HTMLDivElement.prototype.getBoundingClientRect = getBoundingClientRect;

      const step = 10;
      render(<Slider step={step} />);

      const sliderElement = screen.getByRole('slider');
      expect(sliderElement).toHaveAttribute('style', 'left: 50px;');
      global.document['window'] = original;
    });
  });
  describe('Når title ikke settes', (): void => {
    test('Så rendres ikke title elementet', (): void => {
      const { container } = render(<Slider />);

      const titleElement = container.getElementsByTagName('h3');
      expect(titleElement.length).toBe(0);
    });
  });
  describe('Når optionLeft og optionRight ikke settes', (): void => {
    test('Så rendres ikke options tekst', (): void => {
      const { container } = render(<Slider />);

      const titleElement = container.getElementsByTagName('p');
      expect(titleElement.length).toBe(0);
    });
  });
  describe('Når disabled settes til true', (): void => {
    test('Så deaktiveres slideren', (): void => {
      render(<Slider disabled={true} />);

      const sliderElement = screen.getByRole('slider');
      expect(sliderElement.className).toBe('slider__trigger slider__trigger--disabled');
    });
  });
  describe('Når onChange settes', (): void => {
    test('Så kalles funksjonen som den ble satt til', (): void => {
      const mockFunction = jest.fn();
      render(<Slider onChange={mockFunction} />);

      const sliderElement = screen.getByRole('slider');
      fireEvent(
        sliderElement,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );

      expect(mockFunction).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når keyDown kalles', (): void => {
    test('Så settes verdi riktig ved bevegelse av slider triggeren', () => {
      const original = global.document['window'];

      const getBoundingClientRect = jest
        .fn()
        .mockReturnValueOnce({ left: 100, right: 900 })
        .mockReturnValueOnce({ left: 150, right: 850 });
      window.HTMLDivElement.prototype.getBoundingClientRect = getBoundingClientRect;

      const step = 10;
      render(<Slider step={step} />);

      const sliderElement = screen.getByRole('slider');
      expect(sliderElement).toHaveAttribute('aria-valuenow', '50');
      expect(sliderElement).toHaveAttribute('style', 'left: 50px;');

      fireEvent.keyDown(sliderElement, { key: 'ArrowRight', code: 'ArrowRight' });
      expect(sliderElement).toHaveAttribute('aria-valuenow', '60');
      expect(sliderElement).toHaveAttribute('style', 'left: 60px;');

      fireEvent.keyDown(sliderElement, { key: 'ArrowLeft', code: 'ArrowLeft' });
      fireEvent.keyDown(sliderElement, { key: 'ArrowLeft', code: 'ArrowLeft' });
      expect(sliderElement).toHaveAttribute('aria-valuenow', '40');
      expect(sliderElement).toHaveAttribute('style', 'left: 40px;');
      global.document['window'] = original;
    });
  });
});
