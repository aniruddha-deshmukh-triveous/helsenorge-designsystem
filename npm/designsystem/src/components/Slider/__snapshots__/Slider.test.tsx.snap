// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Slider skal vises Når slider rendres Så vises slider riktig 1`] = `
<div>
  <div
    class="slider"
  >
    <h3
      class="sc-AxjAm jzYRRO"
    >
      This is a title
    </h3>
    <div
      class="slider__track-wrapper"
      data-testid="tracker"
    >
      <div
        class="slider__track"
      />
      <div
        aria-label="This is a title left right "
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="50"
        class="slider__trigger"
        role="slider"
        style="left: 0px;"
        tabindex="0"
      >
        <div
          class="slider__trigger-inner"
        />
      </div>
    </div>
    <span
      class="slider__options"
    >
      <p
        aria-hidden="true"
        class="slider__text"
      >
        left
      </p>
      <p
        aria-hidden="true"
        class="slider__text"
      >
        right
      </p>
    </span>
  </div>
</div>
`;
